### ISSUE TITLE
 latest version autogluon==0.0.5  F1 macro score dropped significantly

### ISSUE THREAD

**Issue by quuhua911** [awslabs/autogluon#274]
I was working with the latest version autogluon==0.0.5 on a multi category table prediction problem and found that the run time was significantly shorter and the F1 macro score dropped significantly compared to the previous version.

dir = './agModels-predictClass' # specifies folder where to store trained models
predictor = task.fit(train_data=train_data, eval_metric ="f1_macro",hyperparameters ={ 'GBM': {'num_boost_round': 10000},'custom': ['GBM']},
                     num_trials=60,nthreads_per_trial=30,auto_stack=True,verbosity=2,label=label_column,problem_type="multiclass", output_directory=dir)

Beginning AutoGluon training ...
AutoGluon will save models to ./agModels-predictClass/
Train Data Rows:    322757
Train Data Columns: 56
Preprocessing data ...
Feature Generator processed 322757 data points with 55 features
Original Features:
	object features: 1
	float features: 50
	int features: 4
Generated Features:
	int features: 0
All Features:
	object features: 1
	float features: 50
	int features: 4
	Data preprocessing and feature engineering runtime = 3.54s ...
AutoGluon will gauge predictive performance using evaluation metric: f1_macro
To change this, specify the eval_metric argument of fit()
AutoGluon will early stop models using evaluation metric: f1_macro
Fitting model: LightGBMClassifier_STACKER_l0 ...
	0.8348	 = Validation f1_macro score
	181.67s	 = Training runtime
	0.1s	 = Validation runtime
Fitting model: LightGBMClassifierCustom_STACKER_l0 ...
	0.187	 = Validation f1_macro score
	42.73s	 = Training runtime
	0.09s	 = Validation runtime
Fitting model: weighted_ensemble_k0_l1 ...
	0.8348	 = Validation f1_macro score
	40.42s	 = Training runtime
	0.14s	 = Validation runtime
Fitting model: LightGBMClassifier_STACKER_l1 ...
	0.8413	 = Validation f1_macro score
	105.52s	 = Training runtime
	0.15s	 = Validation runtime
Fitting model: LightGBMClassifierCustom_STACKER_l1 ...
	0.187	 = Validation f1_macro score
	45.69s	 = Training runtime
	0.09s	 = Validation runtime
Fitting model: weighted_ensemble_k0_l2 ...
	0.8413	 = Validation f1_macro score
	38.41s	 = Training runtime
	0.1s	 = Validation runtime
AutoGluon training complete, total runtime = 459.26s ...

**Comment by Innixma** [comment]
Thanks for submitting this issue!

F1 score has a defect in it in the current version, specifically because we are not adjusting the threshold (Currently fixed at 0.5), although I think this problem should have also been present in earlier versions too.

Also, you are specifying HPO num_trials but you did not enable HPO: hyperparameter_tune=True, meaning it will not do anything.

A workaround is for you to get the pred_proba output of Predictor and manually threshold. A full solution involves thresholding the result internally to maximize validation score, but it is somewhat more involved, and we will look into fixing this in future.

Best,
Nick

**Comment by Innixma** [comment]
I just noticed you are actually doing multiclass classification. Could you provide the logs for your run on the older version of AutoGluon where you got a better score?

**Comment by quuhua911** [comment]
sorry，The logs have been erased along with older version of AutoGluon, so there's no way to recreate them. But Thank you so much for your quick response and your advice.

**Comment by Innixma** [comment]
Can you try:

`pip install autogluon==0.0.4`

and re-running the above code to see the output? One potential reason is because the early stopping rounds were decreased in 0.0.5 for large datasets (previously 150, now 10), meaning it is possible the model is now stopping before it reaches its best score. If this is the case, we can look to update the minimum early stopping rounds to correct for your problem.

**Comment by quuhua911** [comment]
autogluon==0.0.4:
dir = './agModels-predictClass' # specifies folder where to store trained models
predictor = task.fit(train_data=train_data, eval_metric ="f1_macro",hyperparameters ={ 'GBM': {'num_boost_round': 10000},'custom': ['GBM']},
                     num_trials=600,nthreads_per_trial=40,auto_stack=True,verbosity=2,label=label_column,problem_type="multiclass", output_directory=dir)
**Here's the log: As the run time increases, the F 1macro score improves**
Beginning AutoGluon training ...
AutoGluon will save models to ./agModels-predictClass/
Preprocessing data ...
	Data preprocessing and feature engineering runtime = 2.21s ...
AutoGluon will gauge predictive performance using evaluation metric: f1_macro
To change this, specify the eval_metric argument of fit()
Fitting model: LightGBMClassifier_STACKER_l0 ...
[1000]	train_set's multi_error: 0.000616221	train_set's multi_logloss: 0.00923461	train_set's f1_macro: 0.997808	valid_set's multi_error: 0.0234842	valid_set's multi_logloss: 0.100152	valid_set's f1_macro: 0.861175
[1000]	train_set's multi_error: 0.000564581	train_set's multi_logloss: 0.00903674	train_set's f1_macro: 0.997956	valid_set's multi_error: 0.025158	valid_set's multi_logloss: 0.10885	valid_set's f1_macro: 0.844527
[1000]	train_set's multi_error: 0.000488841	train_set's multi_logloss: 0.00889008	train_set's f1_macro: 0.99819	valid_set's multi_error: 0.0258412	valid_set's multi_logloss: 0.116858	valid_set's f1_macro: 0.839764
	4015.51s	 = Training runtime
	0.8496	 = Validation f1_macro score
Fitting model: LightGBMClassifierCustom_STACKER_l0 ...
[1000]	train_set's multi_error: 3.4426e-06	train_set's multi_logloss: 0.00448455	train_set's f1_macro: 0.999983	valid_set's multi_error: 0.0228322	valid_set's multi_logloss: 0.104148	valid_set's f1_macro: 0.868203
[1000]	train_set's multi_error: 3.44257e-06	train_set's multi_logloss: 0.00444602	train_set's f1_macro: 0.999983	valid_set's multi_error: 0.0248792	valid_set's multi_logloss: 0.109761	valid_set's f1_macro: 0.845841
[1000]	train_set's multi_error: 3.44255e-06	train_set's multi_logloss: 0.00444033	train_set's f1_macro: 0.999983	valid_set's multi_error: 0.0243842	valid_set's multi_logloss: 0.111069	valid_set's f1_macro: 0.858558
[1000]	train_set's multi_error: 0	train_set's multi_logloss: 0.00439193	train_set's f1_macro: 1	valid_set's multi_error: 0.0248187	valid_set's multi_logloss: 0.108757	valid_set's f1_macro: 0.856879
	5482.8s	 = Training runtime
	0.8566	 = Validation f1_macro score
Fitting model: weighted_ensemble_k0_l1 ...
	24.96s	 = Training runtime
	0.8566	 = Validation f1_macro score
Fitting model: LightGBMClassifier_STACKER_l1 ...
	782.81s	 = Training runtime
	0.8581	 = Validation f1_macro score
Fitting model: LightGBMClassifierCustom_STACKER_l1 ...
[1000]	train_set's multi_error: 0	train_set's multi_logloss: 0.00373128	train_set's f1_macro: 1	valid_set's multi_error: 0.0237948	valid_set's multi_logloss: 0.116615	valid_set's f1_macro: 0.863432
[1000]	train_set's multi_error: 6.88508e-06	train_set's multi_logloss: 0.00373781	train_set's f1_macro: 0.999973	valid_set's multi_error: 0.024385	valid_set's multi_logloss: 0.119059	valid_set's f1_macro: 0.859021
	5594.68s	 = Training runtime
	0.8623	 = Validation f1_macro score
Fitting model: weighted_ensemble_k0_l2 ...
	24.35s	 = Training runtime
	0.8623	 = Validation f1_macro score
AutoGluon training complete, total runtime = 15928.15s ...
1
test_data = task.Dataset(all_data_test)
2
y_test = test_data[label_column]  # values to predict
3
test_data_nolab = test_data.drop(labels=[label_column],axis=1) # delete label column to prove we're not cheating
4
#print(test_data_nolab.head())
1
predictor = task.load(dir) # unnecessary, just demonstrates how to load previously-trained predictor from file
2
​
3
y_pred = predictor.predict(test_data_nolab)
4
print("Predictions:  ", y_pred)

**Comment by quuhua911** [comment]
> Can you try:
> 
> `pip install autogluon==0.0.4`
> 
> and re-running the above code to see the output? One potential reason is because the early stopping rounds were decreased in 0.0.5 for large datasets (previously 150, now 10), meaning it is possible the model is now stopping before it reaches its best score. If this is the case, we can look to update the minimum early stopping rounds to correct for your problem.

I have enough computing power and memory, and I'm after good model results, so...

**Comment by Innixma** [comment]
I see. Thanks for the logs! To be clear: Reducing early stopping rounds is not intended to get worse results. It is actually to combat overfitting and speed up results. Previously, we would do 150 rounds of early stopping, which would tend to take an unreasonable amount of time for large datasets where the best score was found around iterations 1-10: Meaning we were taking >15x the time that we could have taken.

However, some metrics, particularly F1 Score, are difficult to properly early stop as they reach their maximums at different times than log loss.

This is something we didn't properly account for when limiting the early stopping rounds on large datasets, and will look to alter going forward, but it is a complex issue: Raising the early stopping rounds may help your score, but hurt others.

Also: **I would recommend including CAT in your models** if you want the best performance, it works very well with GBM.

The plan: Add option in hyperparameters to specify early stopping rounds.

**Comment by Innixma** [comment]
#714 is tracking adding more user-options for early stopping control.
